tower/src/builder/mod.rs

```diff
@@ -0,0 +1,198 @@
+/// #        future::ok(())
+/// #    }
+/// # }
+/// ServiceBuilder::new()
+///     .chain(InFlightLimitLayer::new(5))
+///     .build_maker(MyMakeService);
+/// ```
+///
+/// A `Service` stack with a single layer:
+///
+/// ```
+/// # extern crate tower;
+/// # extern crate tower_in_flight_limit;
+/// # extern crate futures;
+/// # extern crate void;
+/// # use void::Void;
+/// # use tower::Service;
+/// # use tower::builder::ServiceBuilder;
+/// # use tower_in_flight_limit::InFlightLimitLayer;
+/// # use futures::{Poll, future::{self, FutureResult}};
+/// # #[derive(Debug)]
+/// # struct MyService;
+/// # impl Service<()> for MyService {
+/// #    type Response = ();
+/// #    type Error = Void;
+/// #    type Future = FutureResult<Self::Response, Self::Error>;
+/// #    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+/// #        Ok(().into())
+/// #    }
+/// #    fn call(&mut self, _: ()) -> Self::Future {
+/// #        future::ok(())
+/// #    }
+/// # }
+/// ServiceBuilder::new()
+///     .chain(InFlightLimitLayer::new(5))
+///     .build_service(MyService);
+/// ```
+///
+/// A `Service` stack with _multiple_ layers that contain rate limiting, in-flight request limits,
+/// and a channel-backed, clonable `Service`:
+///
+/// ```
+/// # extern crate tower;
+/// # extern crate tower_in_flight_limit;
+/// # extern crate tower_buffer;
+/// # extern crate tower_rate_limit;
+/// # extern crate futures;
+/// # extern crate void;
+/// # use void::Void;
+/// # use tower::Service;
+/// # use tower::builder::ServiceBuilder;
+/// # use tower_in_flight_limit::InFlightLimitLayer;
+/// # use tower_buffer::BufferLayer;
+/// # use tower_rate_limit::RateLimitLayer;
+/// # use std::time::Duration;
+/// # use futures::{Poll, future::{self, FutureResult}};
+/// # #[derive(Debug)]
+/// # struct MyService;
+/// # impl Service<()> for MyService {
+/// #    type Response = ();
+/// #    type Error = Void;
+/// #    type Future = FutureResult<Self::Response, Self::Error>;
+/// #    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+/// #        Ok(().into())
+/// #    }
+/// #    fn call(&mut self, _: ()) -> Self::Future {
+/// #        future::ok(())
+/// #    }
+/// # }
+/// ServiceBuilder::new()
+///     .chain(BufferLayer::new(5))
+///     .chain(InFlightLimitLayer::new(5))
+///     .chain(RateLimitLayer::new(5, Duration::from_secs(1)))
+///     .build_service(MyService);
+/// ```
+#[derive(Debug)]
+pub struct ServiceBuilder<L, S, Request> {
+    layer: L,
+    _pd: PhantomData<(S, Request)>,
+}
+
+impl<S, Request> ServiceBuilder<Identity, S, Request> {
+    /// Create a new `ServiceBuilder` from a `MakeService`.
+    pub fn new() -> Self {
+        ServiceBuilder {
+            layer: Identity::new(),
+            _pd: PhantomData,
+        }
+    }
+}
+
+impl<L, S, Request> ServiceBuilder<L, S, Request> {
+    /// Chain a layer `T` to the `ServiceBuilder`.
+    pub fn chain<T>(self, layer: T) -> ServiceBuilder<Chain<T, L>, S, Request> {
```

Comment:
"chaining" a layer is mixing metaphors. I would vote for naming this fn `layer` and rename any other usage of the "chain" metaphor.